File: engine/julia/cashflow_capital_ecosystem_engine.jl
# Module Z5 (Julia): Top-Down Cash Flow & Capital Ecosystem Engine
# Precision mathematical modeling of cashflow allocation, sinking fund amortizations, and leak litmus tests.

module CashflowCapitalEcosystemEngine

export calculate_ecosystem, sinking_fund_installment, leak_litmus_test

function sinking_fund_installment(target::Float64, periods::Int, buffer::Float64=0.10)
    if periods <= 0
        return target * (1.0 + buffer)
    end
    return (target * (1.0 + buffer)) / Float64(periods)
end

function calculate_ecosystem(income::Float64, fixed::Float64, variable::Float64, sinking::Float64, savings_ratio::Float64=0.20)
    total_essentials = fixed + variable + sinking
    workable = max(0.0, income - total_essentials)
    keep_alloc = workable * savings_ratio
    spend_alloc = max(0.0, workable - keep_alloc)
    
    return (
        total_income = income,
        workable_total = round(workable, digits=2),
        keep_savings = round(keep_alloc, digits=2),
        spend_discretionary = round(spend_alloc, digits=2)
    )
end

function leak_litmus_test(category::String, est_freq::Int, est_spend::Float64, act_freq::Int, act_spend::Float64)
    freq_diff = act_freq - est_freq
    spend_leak = act_spend - est_spend
    return (
        category = category,
        freq_diff = freq_diff,
        monetary_leak = round(spend_leak, digits=2),
        is_leaking = spend_leak > 0.0 || freq_diff > 0
    )
end

end # module

File: engine/rust/src/cashflow_capital_ecosystem_engine.rs
//! Module Z4 (Rust): Top-Down Cash Flow & Capital Ecosystem Engine.
//! High-speed cashflow routing and sinking fund amortization processor.

#[repr(C, align(64))]
#[derive(Debug, Clone, Copy)]
pub struct CashflowCapitalState {
    pub total_income: f32,
    pub fixed_essentials: f32,
    pub variable_essentials: f32,
    pub sinking_funds_total: f32,
    pub workable_total: f32,
    pub keep_savings_allocated: f32,
    pub spend_discretionary: f32,
    pub new_zero_buffer: f32,
    pub intentional_spend_ratio: f32,
    pub weighted_joy_score: f32,
    pub is_leaking_detected: u32,
    pub values_aligned_flag: u32,
    pub _padding: [u8; 16],
}

pub struct CashflowCapitalEcosystemEngine;

impl CashflowCapitalEcosystemEngine {
    pub fn new_state() -> CashflowCapitalState {
        CashflowCapitalState {
            total_income: 0.0,
            fixed_essentials: 0.0,
            variable_essentials: 0.0,
            sinking_funds_total: 0.0,
            workable_total: 0.0,
            keep_savings_allocated: 0.0,
            spend_discretionary: 0.0,
            new_zero_buffer: 100.0,
            intentional_spend_ratio: 1.0,
            weighted_joy_score: 10.0,
            is_leaking_detected: 0,
            values_aligned_flag: 1,
            _padding: [0; 16],
        }
    }

    pub fn compute_ecosystem(
        state: &mut CashflowCapitalState,
        income: f32,
        fixed_costs: f32,
        variable_costs: f32,
        sinking_total: f32,
        savings_ratio: f32,
        new_zero: f32,
    ) {
        state.total_income = income;
        state.fixed_essentials = fixed_costs;
        state.variable_essentials = variable_costs;
        state.sinking_funds_total = sinking_total;
        state.new_zero_buffer = new_zero;

        let total_essentials = fixed_costs + variable_costs + sinking_total;
        state.workable_total = (income - total_essentials).max(0.0);
        state.keep_savings_allocated = state.workable_total * savings_ratio;
        state.spend_discretionary = (state.workable_total - state.keep_savings_allocated).max(0.0);
        state.values_aligned_flag = if state.keep_savings_allocated > 0.0 && state.workable_total > 0.0 { 1 } else { 0 };
    }
}
